src/server.cpp

```diff
@@ -24,95 +22,38 @@ std::string packetTypeToString(int type)
     return "uncnown";
 }
 
-void Server::printPacket(const Packet& p)
-{
-    std::cout << "Packet type: " << packetTypeToString(p.type()) << "\n";
-
-    std::cout << "ID: " << std::to_string(p.id()) << "\n";
-
-    std::cout << "Attributes:\n";
-    for (const auto& ap : p.attributes())
-        std::cout << "\t" << m_dictionaries.attributes().name(ap->type()) << ": " << ap->toString() << "\n";
-
-    for (const auto& ap : p.vendorSpecific())
-    {
-        const std::string vendorName = m_dictionaries.vendorNames().name(ap->vendorId());
-        std::cout << "\tVendor-Specific" << ": " << vendorName << "\n";
-        std::cout << "\t" << m_dictionaries.vendorAttributes().name(vendorName, ap->vendorType()) << ": " << ap->toString() << "\n";
-    }
-}
-
 Server::Server(boost::asio::io_service& io_service, const std::string& secret)
-      : m_socket(io_service, udp::endpoint(udp::v4(), 9999)),
-        m_secret(secret),
-        m_dictionaries("/usr/share/freeradius/dictionary")
+    : m_socket(io_service, udp::endpoint(udp::v4(), 9999)),
+      m_secret(secret),
+      m_dictionaries("/usr/share/freeradius/dictionary")
 {
-    startReceive();
 }
 
-void Server::startReceive()
+void Server::asyncReceive(std::function<void(const error_code& error, const Packet& request)> callback)
 {
     m_socket.async_receive_from(boost::asio::buffer(m_recvBuffer), m_remoteEndpoint,
-        [this](const error_code& error, std::size_t bytes) {handleReceive(error, bytes);});
+       [this, callback](const error_code& error, std::size_t bytes) {handleReceive(error, bytes, callback);});
 }
 
-void Server::handleReceive(const error_code& error, std::size_t bytes)
+void Server::asyncSend(std::function<void(const error_code& ec)> callback, Packet& response)
 {
-    if (error)
-    {
-        std::cout << "Error async_receivw_from: " << error.message() << "\n";
-        return;
-    }
-
-    if (bytes < 20)
-    {
-        std::cout << "Error: request length less than 20 bytes\n";
-        return;
-    }
+    m_socket.async_send_to(boost::asio::buffer(response.makeSendBuffer(m_secret)), m_remoteEndpoint,
+       [this, callback](const error_code& ec, std::size_t /*bytesTransferred*/) {handleSend(ec, callback);});
+}
 
+void Server::handleReceive(const error_code& error, std::size_t bytes, std::function<void(const error_code&, const Packet&)> callback)
+{
     try
     {
-        Packet packet = makeResponse(Packet(m_recvBuffer, bytes, m_secret));
-
-        std::cout << "Response packet\n";
-        printPacket(packet);
-
-        m_socket.async_send_to(boost::asio::buffer(packet.makeSendBuffer(m_secret)), m_remoteEndpoint,
-           [this](const error_code& /*ec*/, std::size_t /*bytesTransferred*/) {handleSend();});
+        const Packet request = Packet(m_recvBuffer, bytes, m_secret);
```

Comment:
You used to check error code and number of bytes before. You can not create a packet if the operation resulted in error, or if we received less than 20 bytes.